### PR TITLE
Fix #3948 - Changing state was also causing change of initial value

### LINKED ISSUE #3948: Formik.dirty is not reacting to nested changes
### Current Behavior

I have an object with three-level nested object
When I change it formik values is changing but formik.dirty is not reacting except if i add/remove elements

### Expected behavior
formik.dirty should react to nested objects changes 

### Reproducible example

https://codesandbox.io/p/sandbox/nested-schema-bug-gy44pw?file=/index.js

### My environment

| Software         | Version(s) |
| ---------------- | ---------- |
| Formik           | 2.4.5 | 
| React            | nextjs 13.2.3 | 
| TypeScript       | no | 
| Browser          | arc 1.27.3| 
| npm/Yarn         | npm | 
| Operating System | macOS 13.5.1|

### PR DESCRIPTION
**Pull request by DeveloperRaj**
Resolves Issue #3948

This addresses the problem of the dirty field not updating when the value of a nested object changes. The root cause of this issue is that the `initialValues` coming from props were directly assigned to the `useRef`, which did not perform a deep copy. Without a deep copy, it was modifying the original `initialValues` props along with the current state. Consequently, when comparing them for equality, the result was true

### CODE REVIEW

File: packages/formik/src/Formik.tsx [hunk 1]
```diff
@@ -29,6 +29,7 @@ import {
   setIn,
   setNestedObjectValues,
 } from './utils';
+import cloneDeep from 'lodash/cloneDeep';
```
Comment (quantizor) [inline packages/formik/src/Formik.tsx:32]: Can you move this import up with the other lodash one and add a changeset please?

Comment (DeveloperRaj) [reply packages/formik/src/Formik.tsx]: Done

**Comment by quantizor** [comment]
@DeveloperRaj thank you, last thing can you add a unit test that verifies the `initialValues` object is not modified?

**Comment by DeveloperRaj** [comment]
> @DeveloperRaj thank you, last thing can you add a unit test that verifies the `initialValues` object is not modified?

@quantizor Thanks for suggestion of test. I have added a new test that checks that initial values are not being modified when we do handleChange or setValues etc. It looks a bit messy but had no option since this bug would repro only happen when we have deeply nested objects otherwise it would work just fine even without my code change.

**Review by quantizor — APPROVED** [review]
Thanks!

**Pull request merged**